refactor recurrent, update examples, add tests

skaae opened the pull request.

skaae commented:
I added the following flags to the recurrent layers
- return_sequence: If set the layer returns the last hidden state. This is useful when you want to make a
  single prediction for a sequence or want to implement an encoder/deocer
- grad_clipping: Grad clipping like Alex Graves
- return_cell: optionally returns the cell. I think this might slightly violate Lasagnes output format, but it is useful if you want to implement and encoder/decoder.

Do you agree on the additional flags?

I would also like to add an option to use the previous output as input in the next timestep. It requires that the user can pass in an output network? It is neccesary for language models etc.  

I'll go over the doc strings again tomorrow and fix typos. When this is merged i'll add an embedding layer + step clipping and write the penn tree example.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -1,55 +1,98 @@
         (n_batch, n_time_steps, n_features_1, n_features_2, ...)
-    '''
-    def __init__(self, input_layer, input_to_hidden, hidden_to_hidden,
+
```
How come you moved the docstring up here?  These parameters are arguments for the `__init__` function, not the `CustomRecurrentLayer`.  Usually, the docstring of the class includes attributes of the class, while the docstring of the `__init__` method includes arguments to the `__init__` function.  See https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#documenting-classes

benanne replied:
Actually, the numpydoc approach is to document the constructor arguments in the class docstring (it actually says so in the document you linked, "The constructor (`__init__`) should also be documented here"). We've been doing it this way for Lasagne as well.

craffel replied:
Oh, you're right!  I guess I've always been doing it wrong.  Thanks.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -1,55 +1,98 @@
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape
+    input_to_hidden : nntools.layers.Layer
```
These should all be changed to lasagne.layers.Layer

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -1,55 +1,98 @@
+        Number of timesteps to include in backpropagated gradient
+        If -1, backpropagate through the entire sequence
+    return_sequence : boolean
+        Specifies if the LSTM should output the hidden state for all positions
```
This isn't the LSTM layer, please change to "recurrent layer".

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -1,55 +1,98 @@
+    return_sequence : boolean
+        Specifies if the LSTM should output the hidden state for all positions
+        in the sequence or only for the last position. If true the output
+        shape is (num_batch, sequence_lengt, num_units) if false the output
```
"sequence_lengt" typo

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -1,55 +1,98 @@
+        shape is (num_batch, num_units).
+    grad_clipping: False or float
+        If float the gradient messages are clipped during the backward pass.
+        See [2]_ (p. 6) for further explanation.
```
I think this reference ([2]_) will be undefined because you define it in the `LSTMLayer` docstring.

skaae replied:
true, i'll add it as a reference.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -133,19 +156,27 @@ def get_output_for(self, input, mask=None, *args, **kwargs):
+                    hid_pre, -self.grad_clipping, self.grad_clipping)
+            return self.nonlinearity(hid_pre)
+
+        def step_masked(layer_input, mask_n, hid_previous):
```
Curious why you changed the variable name to `mask_n`.

skaae replied:
I got a warning saying that it masked the outer mask variable. And i figured that input_dot_w was named input_dot_w_n.

craffel replied:
Makes sense!

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -153,61 +184,78 @@ def step_masked(layer_input, mask, hid_previous):
-    def __init__(self, input_layer, num_units, W_in_to_hid=init.Uniform(),
-                 W_hid_to_hid=init.Uniform(), b=init.Constant(0.),
+
+    Parameters
```
Same comment as above where the params should be in `__init__`, I believe.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -221,25 +269,102 @@ def __init__(self, input_layer, num_units, W_in_to_hid=init.Uniform(),
-           regularization"
-    '''
-    def __init__(self, input_layer, num_units,
+    Parameters
```
Same comment about params in `__init__` docstring.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -564,9 +550,9 @@ def get_output_for(self, input, mask=None, *args, **kwargs):
         # the inputs to the gates before scanning.
         # input is dimshuffled to (n_time_steps, n_batch, n_features)
-        # W_in_to_gates is (n_features, 4*num_units). input_dot_W is then
+        # W_in_to_gates is (n_features, 4*num_units). input_dot_w is then
```
Any particular reason you made the `w` lowercase?

skaae replied:
no, except for a pep8 error/warning that i got rid of. But I can change it to upper_case.

craffel replied:
Ah, right, I wonder why my PEP-8 checker didn't catch that.  Cool.

skaae replied:
All the parameter  (W_something ) names also gives errors/warnings, but uppercase is the convention in Lasagne.

benanne replied:
Really? Are you talking about PEP8? I've never seen it complain about that.

skaae replied:
Its pycharns PEP8. I don't know if it is correct.
![pep8](https://cloud.githubusercontent.com/assets/2623134/7677376/cc4af1ea-fd4b-11e4-8309-12518d9b2db6.png)

f0k replied:
There actually is a flake8 plugin for that: https://pypi.python.org/pypi/pep8-naming
I can't find this convention in PEP8, though: https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments
It only tells which convention to use for functions, methods and variables, but not which one to use for function and method arguments.
I'd say we should keep references to `W` uppercase then.

craffel replied:
In that case, @skaae, can you switch it back to `input_dot_W`/`input_dot_W_n`, assuming you agree that's more clear?

skaae replied:
yes. Can you clearify what notation you prefer :) I don't really have an opinion.

craffel replied:
Do you want to revert/change whichever commit you made the change to the lowercase w, or do you want me to make another commit where I change it back?  The former is much cleaner but I'm not going to force you to do it! :)

skaae replied:
I can do it, I just wasn't sure what version you prefered :) 

Just so we are clear:

`input_dot_W` and `input_dot_W_n` inside scan?

I think we should squash the commits at the end anyway? 

EDIT: I'll add it through a commit with all of your sugesstions.

craffel commented:
Thanks very much for this!  I think I'd prefer all of the interface/docstring updates to be in a separate PR than the new features, but that's just nitpicking, this is fine.  It also would be great if there were multiple commits which atomized the changes, but that's also nitpicking.

For `return_sequences`, what's the use case for that? Is there any reason it's better than just calling `get_output` and slicing the last entry from the resulting Theano expression?

I've added a few other misc comments inline in the diffs.  I can make the changes myself on top of your commits if you prefer.  I haven't taken a thorough look at the tests but they look great!  Thanks again.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -611,42 +602,59 @@ def step(input_dot_W_n, cell_previous, hid_previous):
+                    cell_out = cell_out[:, ::-1, :]
+            else:
+                cell_out = cell_out[-1]
+            output = [cell_out, output]
```
@f0k and @benanne - what do you think about this, where the output of the layer is a two-element list with two Theano `TensorVariable`s?  I think @skaae is doing this so that you can make an encoder/decoder RNN using the LSTM, but I'm not totally sure what implications there are to having a layer with a list output.

Another way to do this would be to create a separate layer which just performs the gating and not the output linear mix/nonlinearity, and then if you want to have a normal `LSTMLayer` you can add the `DenseLayer` yourself.  We can also, of course, still provide an `LSTMLayer` which basically would just construct the gating layer itself and then apply the dense transformation to the cell output itself.  It's worth thinking about this in the context of the `GRULayer` too - all of these things should probably share a lot of code.

benanne replied:
At the moment I don't think there is anything in Lasagne that relies on a layer's output being a Theano expression - except of course other layers. So that would mean it's not necessarily a problem to return other data types from `get_output_for()`.

However, you would then need a special set of layers that can take lists of expressions as input. (it would not be possible to stack a dense layer on top of this for example - it doesn't make sense). That sounds like it could get complicated. Maybe there is a better solution.

I don't fully understand why this is necessary to be able to build an encoder/decoder model, though. Why do you need the cell values for that? I don't really understand the alternatives you proposed either, but that's probably down to not knowing enough about LSTMs. I should go and read up first and then come back :)

skaae replied:
_For return_sequences, what's the use case for that?_
I in general if you run the RNN over some input sequence and want a single prediction. I used it for protein analysis, but you might use it for twitter sentiment analysis or as the very artificial MNIST example in here: http://arxiv.org/abs/1504.00941 .

If we want to move the slicing outputside of the recurrent layers we can instead create a slicing layer. Then the one would still have the possibility to create encoder/decoders? 

_However, you would then need a special set of layers that can take lists of expressions as input. (it would not be possible to stack a dense layer on top of this for example - it doesn't make sense). That sounds like it could get complicated. Maybe there is a better solution._

To create an encoder/decoder as it is constructed in http://arxiv.org/abs/1409.3215 you need to initialize the decoder with the last hidden and cell states from the encoder. The solution with returning two layers is a hack. The user would need to write a decoder that handels the list input.
I tested this a while ago and it works fine.    

The GRU layer does not have the problem with an additional cell value :) 

_Another way to do this would be to create a separate layer which just performs the gating and not the output linear mix/nonlinearity, and then if you want to have a normal LSTMLayer you can add the DenseLayer yourself. We can also, of course, still provide an LSTMLayer which basically would just construct the gating layer itself and then apply the dense transformation to the cell output itself. It's worth thinking about this in the context of the GRULayer too - all of these things should probably share a lot of code._

I dont think that would solve the problem with the need to return both cell and hidden values.

Should i add the outputnetwork output to allow for language models?

f0k replied:
> If we want to move the slicing outputside of the recurrent layers we can instead create a slicing layer. 

This sounds like the more general solution.

> what do you think about this, where the output of the layer is a two-element list with two Theano `TensorVariables`?

So far we didn't have a use case for a layer that returns multiple outputs. Here it seems the operation cannot simply be expressed as two single-output layers because both outputs come from the same `scan` op.
So maybe we should just add support for multiple-output layers, and the natural way to do so would be returning lists/tuples of outputs in `get_output_for` and `get_output_shape_for`. There could be other use cases for multiple-output layers. To allow stacking ordinary layers on top, we could have a `DemuxLayer` that just returns one of its multiple inputs. However, I think we shouldn't allow layers that either produce multiple outputs or not -- would they have a `self.output_shape` or a `self.output_shapes` property?

craffel replied:
> > If we want to move the slicing outputside of the recurrent layers we can instead create a slicing layer.
> > This sounds like the more general solution.

Agreed.  I think we should create a slicing layer and not have a `return_sequences` kwarg.

> I dont think that would solve the problem with the need to return both cell and hidden values.

I think it actually would (I may not have explained it well), but after considering it more I think it wouldn't allow for peephole connections, so it's not really a viable option.

> So maybe we should just add support for multiple-output layers, and the natural way to do so would be returning lists/tuples of outputs in get_output_for and get_output_shape_for. There could be other use cases for multiple-output layers. To allow stacking ordinary layers on top, we could have a DemuxLayer that just returns one of its multiple inputs. However, I think we shouldn't allow layers that either produce multiple outputs or not -- would they have a self.output_shape or a self.output_shapes property?

Agreed here in general.  I think in this particular case `self.output_shape` refers to the shape for both the cell and LSTM layer output, isn't that the case @skaae? But I could see that not being true for some other layer with multiple outputs.

@skaae, I'd like to merge your docstring audits/updates/tests/etc, but I don't think we've resolved discussion on the "output cell"/"layer with multiple outputs" question.  Can you move it to another PR so I can merge?

Thanks again.

skaae replied:
I think that until more use cases with multiple outputs turn up we can just do with the return_cell hack. But we can discuss it in a seperate issue.
`self.output_shape` refers to the shape of both layers.

If you agree i'll do the following:
1. remove the `return_sequences` flag and create PR for a slicing layer. 
2. remove `return_cell` and create a new issue.

skaae commented:
I'll add your suggestions and fix a few more things.

skaae commented:
Removed `return_cell` and `return_sequence` args and modified the tests.

I opened an issue for returning cell values: https://github.com/craffel/nntools/issues/30

EDIT: Remember to close https://github.com/craffel/nntools/issues/17 when this is merged.

craffel commented:
Thanks for this, I'll do a final audit/code review later!

> EDIT: Remember to close #17 when this is merged.

I'm guessing you meant https://github.com/Lasagne/Lasagne/issues/17 ?  That's for merging this fork into the master branch; merging this PR would merge your fork into this fork, so we aren't there quite yet.

skaae commented:
No I meant https://github.com/craffel/nntools/issues/17

I fixed it with a6c2e8afb68c25f9f9596fc2770982110aab55f0

craffel commented:
Oops, got it, didn't realize you were fixing that too.

skaae commented:
There is a problem with all the layers if you use batchsize 1.

Heres a test that fails:

``` Python
def test_batchsize_one():
    theano.config.compute_test_value = 'raise'
    num_batch, seq_len, n_features = 1, 3, 10
    num_units = 6
    x_in = np.random.random((num_batch, seq_len, n_features)).astype('float32')
    x = T.tensor3()
    x.tag.test_value = x_in
    l_inp = InputLayer((num_batch, seq_len, n_features))

    mask_in = np.random.random((num_batch, seq_len)).astype('float32')
    mask = T.matrix()
    mask.tag.test_value = mask_in
    l_lstm = RecurrentLayer(l_inp, num_units=num_units)
    l_out = helper.get_output(l_lstm, x, mask=mask)
    f_lstm = theano.function([x, mask], l_out)
    f_out = f_lstm(x_in, mask_in)
    theano.config.compute_test_value = 'off'
    assert f_out.shape == (num_batch, seq_len, num_units)
```

Turns out you can fix it with

``` Python
        if self.n_batch > 1:
            hid_init = T.dot(T.ones((self.n_batch, 1)), self.hid_init)
        else:
            hid_init = self.hid_init
        hid_out = theano.scan(
            step_fun,
            sequences=sequences,
            go_backwards=self.backwards,
            outputs_info=[hid_init],
            truncate_gradient=self.gradient_steps)[0]
```

I checked and it doesn't change any shapes/ datatypes, so it might be a bug in Theano.

Unless you have a good explanation is think we should just insert the work around and report the bug. 

It happens on both GPU and CPU

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -58,70 +97,47 @@ def __init__(self, input_layer, input_to_hidden, hidden_to_hidden,
+        params = (helper.get_all_params(self.input_to_hidden, **tags) +
+                  helper.get_all_params(self.hidden_to_hidden, **tags))
 
         if self.learn_init:
```
If self.hid_init has the trainable flag set (see line 105), can’t we omit this part now?

f0k replied:
You actually _have to_ omit the `if` clause. Instead, you need to call the super class `get_params`. That's what I've been preaching: https://github.com/Lasagne/Lasagne/pull/228#issuecomment-102833331
/edit: For clarification, the current code does the wrong thing when called with `trainable=False`: In this case, it should return the initialization parameter when `learn_init=False`, and should not return it when `learn_init=True`. Similarly, it does the wrong thing when called with `tags={}`: It should return the initialization parameter unconditionally, but it only returns it when `learn_init=True`. Using the super class `get_params` will take care of all this, because the initialization parameter was tagged correctly in the constructor already.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -153,61 +177,74 @@ def step_masked(layer_input, mask, hid_previous):
-                             go_backwards=self.backwards,
-                             outputs_info=[self.hid_init],
-                             truncate_gradient=self.gradient_steps)[0]
+        # repeat hid_init to batch size
```
I think computing a dot product here is more expensive than using the broadcastable attribute of hid_init.  Can we do that instead?

f0k replied:
I'm not sure if that would work with `scan`. You want the output to be of shape `(batchsize, ...)`, so probably you cannot provide a tensor of shape `(1, ...)` in the `outputs_info` argument (even if it is broadcastable in the first dimension). It's worth trying, though. Instead of the dot product, you'd just do `theano.tensor.shape_padleft(self.hid_init, 1)`.

/edit: Another alternative to the dot product is `theano.tensor.tile(theano.tensor.shape_padleft(self.hit_init, 1), (batchsize, 1))`, or something with `theano.tensor.extra_ops.repeat`. Not sure if that would be any more efficient, though.

/edit: Ah, my examples all assume that `self.hid_init` is a vector, not a matrix that already has a `1` in the first dimension.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -611,42 +576,49 @@ def step(input_dot_W_n, cell_previous, hid_previous):
             step_fun = step
 
+        # repeat cell and hid init to batch size
+        ones = T.ones((self.num_batch, 1))
```
Same comment here as above.

skaae replied:
I dont know i you can do that because the hid_init needs to be the same size as hid inside scan?

Did you see the possible bug with batch size 1 

EDIT: somehow missed your last comment

craffel commented:
> There is a problem with all the layers if you use batchsize 1.
> ...

I actually think it'd be better to use `broadcastable` instead of dotting against ones (see the diff comment I just posted), so maybe we can avoid this completely.

Otherwise, looks good to me!

f0k commented on lasagne/layers/recurrent.py:
```diff
@@ -58,70 +97,44 @@ def __init__(self, input_layer, input_to_hidden, hidden_to_hidden,
         # Initialize hidden state
-        self.hid_init = self.create_param(hid_init, (n_batch, self.num_units))
+        self.hid_init = self.add_param(hid_init, (1, self.num_units),
+                                       trainable=learn_init, name="hid_init")
```
Note that this parameter is now tagged as `regularizable`, because that's the default. Is this intended? I.e., when you train an RNN with L2 weight decay and with `learn_init=True`, do you want to constrain the norm of the initialization vector or not? I'm not that familiar with the RNN literature, but my gut feeling would be that you don't regularize it. The initialization vector should provide a neutral hidden state to start from, and the hidden state of an RNN includes an unregularized bias vector, so at least for ReLUs, it could be arbitrarily large.
You could also set `regularizable=False` and `init_state=True`, so people can choose whether or not to regularize them by providing appropriate tags in the `get_all_params()` calls.

skaae replied:
Yes. I forgot. The LSTM init is set to regularizable=False

skaae commented:
I think we should somehow allow the user to specify the state of hid_init and cell_init. 

With the current code there is no way that you can process sequences longer than seq_len correctly. 

We could allow for `hid_init` and `cell_init` to be symbolic variables?

skaae commented:
Sorry for spamming this thread.

I tested broadcasting the inits. It does not work. 

I have reproduced http://arxiv.org/abs/1409.2329 with a slightly modified code available in https://github.com/skaae/nntools/blob/combine/examples/penntree.py

I made the following modifications to the code:
1. Allow for inits to be symbolic. (the implementation can be cleaned up, but it works)
2. add `get_hidden_state` method that returns a handle to the cell and hid outputs from the LSTM

f0k commented:
> add `get_hidden_state` method that returns a handle to the cell and hid outputs from the LSTM

That's not really needed. It's totally fine for people to access the attributes directly (just like it's fine to access `DenseLayer.W`). We may want to document such attributes for all layers, currently we only do that for the convolutional layers.

skaae commented:
Yeah, dont know why i assumed they where private.

skaae commented:
Unless there is any comments i'll add support for `hid_init` and `cell_init` being `TensorVariables` tomorrow. 

Then recurrent.py is fine with me.

f0k commented:
> Unless there is any comments i'll add support for `hid_init` and `cell_init` being `TensorVariables` tomorrow.

That's a tricky thing. When you do that, they can't be parameters any longer, because `get_all_param_values` assumes all parameters are shared variables. If we want to change that, we need to discuss that in #11. (That would also be useful for auto-encoders with tied weights, for example.) What would be your use case?

skaae commented:
I have a usecase in https://github.com/skaae/nntools/blob/combine/examples/penntree.py

When you process a word corpus the sequences are very long. To work around this you proces them in chunks. You then intiaize each chunk with the previous chunks last hidden state. This ofcourse requires that the data is reorderede appropriately.  Reorder in penntree.py does the reordering. 

I was think that we should net initialize hid_init and cell_init if they a tensorvariable is given.

skaae commented:
Update with
- Allows initial values to be Tensors. This does not change the layer behavior except that hid_init and cell_init are not initialized when inits are given as tensors.
- Removes a bug if the input was larger than 3D
- adds cell and hid outputs as self variables so they can be referenced
  from the outside
- Add tests to bring coverage near 100%. The last lines are only
  nonlinearity set to identity if given as none

skaae commented:
This PR has gotten a bit of of hand :) To sum up what i did:
1. `hid_init` and `cell_init` accepts TensorVariables. `self.hid_init` and `self.cell_init` are not added as params if TensorVariables are passed in. 
2. hid_out and cell_out are added as self variables to allow them to be referenced from outside the layer.
3. Removes a bug when the input dimension is larger than 3
4. Use the `strict=True` for `LSTMLayer`. This gives a nice speedup of around 20%. Updating theano and using strict I went from 4400 words per second to 7100 words per second. You can't use strict for the recurrent layers, unless we write a dedicated vanilla RNN layer. (@f0k had a workaround for this, see below )
5. Fixed https://github.com/craffel/nntools/issues/17
6. Added tests. The coverage is 97%. The remaining lines are nonlinearities being set to identity. 
7. Adds grad_clipping as argument. This optionally clips the backward messages.

I just reran the small Penn Tree experiment from http://arxiv.org/abs/1409.2329. The code reproduces his results. 

Edit: Checked https://github.com/craffel/nntools/issues/11, it works. 

The only problems is the batchsize 1 problem that i'll try to reproduce without Lasagne and post as a bug in Theano.

f0k commented:
Sounds very promising!

> Use the `strict=True` for `LSTMLayer`. [...] You can't use strict for the recurrent layers

How come? Can't you automatically infer which shared variables are going to be used in the scan function, via `get_params()`?

> The code reproduces his results.

Out of curiousity, which of the models did you train (small unregularized, medium regularized, large regularized) and what was the training time in comparison to the paper?

skaae commented:
Dont know. It is mentioned in http://deeplearning.net/software/theano/library/scan.html#using-shared-variables-the-strict-flag and Frederic Bastien also mentioned it in the mailing list. 

**How come? Can't you automatically infer which shared variables are going to be used in the scan function, via get_params()?**
It is `hidden_to_hidden` that is used inside scan. I dont know if its enough to just pass in the variables from that network? I thought you should define the computations with the `non_sequences` that you pass in. 

**Out of curiousity, which of the models did you train (small unregularized, medium regularized, large regularized) and what was the training time in comparison?**

I ran the small model. I ran the same experiment using torch https://github.com/craffel/nntools/blob/master/examples/lstm.py  on my machine. 

I ran at around 4500-5000 words per second. That is without the slicing trick though. The slicing trick has just been introduced (I somehow managed to write a piece of LUA code :)) so torch may still be a little faster, but they are in the same ballpark. The large model runs at 1700 words per second in Theano.

EDIT: I seem to have answered an imaginary question at the begining of the reply....

f0k commented:
> I dont know if its enough to just pass in the variables from that network? I thought you should define the computations with the `non_sequences` that you pass in.

Ah, now I see. You're right. That would require some tricks, but it's possible. First you'd need to get out the parameters with `get_all_params()` and pass them as `non_sequences`. The scan function (i.e., the function you pass as `fn=`) would take `*args` to catch them again. Inside the scan function, you need to define the computation in terms of the parameters you caught instead of the parameters linked to the network. That can be done by calling `get_output()` as usual, but then using `theano.clone` to replace what you get with `get_params()` with the parameters you caught.

Something like this:

``` python
def step_function(..., *args):
    ...
    original_output = get_output(self.hidden_to_hidden(...))
    original_params = get_all_params(self.hidden_to_hidden)
    new_params = args
    new_output = theano.clone(original_output, replace=dict(zip(original_params, new_params)))
    ...

... = theano.scan(..., fn=step_function, non_sequences=get_all_params(self.hidden_to_hidden), strict=True)
```

/edit: `strict=True` requires that for all layer implementations, `get_params()` returns all shared variables participating in the forward pass. We currently do this, and it's probably good to promote this. If we don't want to enforce this, we can also set `strict=False` -- if I understand correctly, this does not affect performance, it just adds a check of whether some performance is possibly sacrificed.

skaae commented:
How do you know all this stuff? I'll give it a try.

skaae commented:
Using theano.clone seems to work. This also somehow adresses 

https://github.com/craffel/nntools/issues/20
https://github.com/craffel/nntools/pull/25

f0k commented:
> Using theano.clone seems to work.

Cool!

> How do you know all this stuff?

I'm near the end of my PhD :)

skaae commented:
@craffel Can you review the code and tell what is missing before this is ready to be merged?

craffel commented:
@skaae thanks again for all of this, and sorry for being MIA, I've been traveling and without internet.  I will re-review everything and try to merge later today.

skaae commented:
No worries :) I have one more suggestion. I have observed that it's often faster to unroll scan with a for
loop instead of using scan. 

In a protein example i saw batch times of ~0.095 for scan and 0.065 for unrolled graph. In more complicated graphs i have seen way worse performance from scan.
The graph was unrolled for 25 steps. I suspect that scan might be better if the graph is unrolled longer.

I have this code that is a drop in for scan

``` Python
        if not self.unroll_graph:
            cell_out, hid_out = theano.scan(
                step_fun,
                sequences=sequences,
                outputs_info=[cell_init, hid_init],
                go_backwards=self.backwards,
                truncate_gradient=self.gradient_steps,
                non_sequences=non_seqs,
                strict=True)[0]
        else:
            counter = range(self.input_shape[1])
            if self.backwards:
                counter = counter[::-1]

            output = []
            init = [cell_init, hid_init]
            for i in counter:
                inp = [input_dot_w[i]] + init + non_seqs
                output.append(step(*inp))
                init = output[-1]

            n_out = len(output[0])
            output_scan = []
            for i in range(n_out):
                l = map(lambda x: x[i], output)
                output_scan.append(T.stack(*l))
            cell_out, hid_out = output_scan
```

I think we should add a `unroll_graph` argument. Then the user can check what is better.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -265,77 +414,13 @@ def __init__(self, input_layer, num_units,
+                 gradient_steps=-1,
+                 grad_clipping=False):
+
+        # todo: add subnetwork option.
```
What is this note about?

skaae replied:
I wanted to add an option for a subnetwork. The idea was that the hidden states from the lstm is passed through the subnetwork and used as additional input to the LSTM in the next timestep. I think it is commonly used in decoders. But we can remove it for now.

craffel commented:
>  I have observed that it's often faster to unroll scan with a for loop instead of using scan.

I could go for that.  I'd be suspicious it would end up using more memory because it's caching the results somehow, so if you don't mind verifying that the memory usage is the same, I think it'd be worth including.  Even if it was a larger memory footprint it still would probably be worth including.  I'd suggest we do that in a different PR, this one addresses a lot already!

Just added a few more comments.  One other thing - running the examples now gives a bunch of warnings:

```
UserWarning: The uniform initializer no longer uses Glorot et al.'s approach to determine the bounds, but defaults to the range (-0.01, 0.01) instead.
```

This one is important; we should decide what we want the default initializer to be.  I'm OK with non-Glorot initialization, I haven't seen people do that.  AFAIK there's no widely agreed-upon "best" initialization range for RNNs, although a Normal(0, .01) seems OK.  We could also be progressive and use the identity-matrix-style initialization proposed in http://arxiv.org/pdf/1504.00941.pdf  It's worth noting that with the new initialization, the RNN example no longer really works, so at the very least we should use something which works there.

```
UserWarning: get_all_layers() has been changed to return layers in topological order. The former implementation is still available as get_all_layers_old(), but will be removed before the first release of Lasagne.
```

OK to ignore this one.

```
Warning: In the strict mode, all neccessary shared variables must be passed as a part of non_sequences
```

It looks like you added `strict=True` to the scan calls, which I am in support of, but we should make sure we are passing all necessary shared variables then!

skaae commented:
I think the glorot warning will be removed at Lasagne release. I think unifom 0.1 or 0.05 is a reasonable default. 

`Warning: In the strict mode, all neccessary shared variables must be passed as a part of non_sequences`
I'm not sure if its just information or a warning that we need to address? In the scan doc it says:

`strict – If true, all the shared variables used in fn must be provided as a part of non_sequences or sequences`

I'll post a question on theano-users.

skaae commented:
I looked the source and the warning is raised whenever strict is true.

craffel commented:
Right, but isn't the point of setting `strict=True` to have it raise this warning exactly?  Because not passing in all of the used shared variables makes it slower, or something?  I guess a better question to ask is, if that wasn't the reason, how come you added `strict=True` in the first place?

skaae commented:
I checked and it does raise an error if some shared variable is not passed in. The warning is just badly formulated and confusing.

craffel commented:
> The warning is just badly formulated and confusing.

Yeah, seems like it!  Ok, LMK when you've removed the note about subnetwork, and changed back to capital W, and I'll merge.  I can fuss with the parameter initialization myself at a later date.

skaae commented:
nice. Should I squash this into a single commit?

craffel commented:
Up to you!

skaae committed:
refactor recurrent + various improvements

hid_init and cell_init accepts TensorVariables. self.hid_init and
self.cell_init are not added as params if TensorVariables are passed in.

hid_out and cell_out are added as self variables to allow them to be
referenced from outside the layer.

Removes a bug when the input dimension is larger than 3

Use the strict=True for scan

Fixed #17

Added tests. The coverage is 97%. The remaining lines are
nonlinearities being set to identity.

Adds grad_clipping as argument. This optionally clips the backward
messages.

skaae committed:
update examples

skaae commented:
Eh, wait a second, maybe i screwed the update up..I'll fix that now

skaae commented:
Seems to be fine.

craffel commented:
Looks good to me, merging, thanks @skaae!  Feel free to address the other things mentioned in this thread in other issues/PRs.

skaae commented:
I'll add the scan flag.  Is unroll_rnn a good name or maybe that does not explain the difference very well.

I also have a GRU implentation?

craffel commented:
> I'll add the scan flag. Is unroll_rnn a good name or maybe that does not explain the difference very well.

I'd prefer `unroll_steps` or `unroll_sequence` or `unroll_scan`.

> I also have a GRU implentation?

Sounds good, feel free to add!  It'd be nice if there was a way for the GRU and LSTM layers to share code.

craffel commented:
Question about `hid_init` as a `TensorVariable`: In what circumstances would you want `hid_init` to be a regular `TensorVariable` and not a `TensorSharedVariable`?  If there aren't any, why aren't we adding it to each layer's parameter list using `add_param`?  If there are, why don't we modify `add_param` to allow `TensorVariable` types?  I think it makes sense to have `hid_init` added to the layer's parameters when the user passes it.

skaae commented:
I use `hid_init` and `cell_init` as TensorVariables in the penntree example in https://github.com/craffel/nntools/pull/33 and also in this paper: http://arxiv.org/abs/1409.2329.

I used it because i needed to process a text corpus that was too long to unroll the model over the entire sequence. Instead i rearranged the dataset such that position1 in batch1 was the first 25 words, position1 in batch**2** was the 26-50 word etc .... 
In batch**N** postiion1 the last word was M. Then first word in position2 in batch**1** wouldbe M+1. and so on.

This mean that you can use the last hidden state from the previous batch to initialize the next batch. 

I guess you could do it with shared variables instead Tensors.

skaae commented:
I agree with your points about `add_param`. I changed the code because I didn't think that you could do it without TensorsVariables.  I'll check the Penntree example and see if it runs without TensorVariables.

f0k commented:
> If there are, why don't we modify `add_param` to allow `TensorVariable` types?

Because that would break several other things, see https://github.com/Lasagne/Lasagne/issues/11 and issues linked to it, especially https://github.com/Lasagne/Lasagne/issues/83. While it would definitely be useful, it needs some careful thought.

skaae commented:
After a little thought i know what the problem was :)

When hid_init/cell_init is shared we want it to be of shape `(1, number_of_units_rnn)`. We then repeat that vector to match batch size. 

When you want to do as in the penn tree example then you need to init with a seperate value for each sample in the batch, i.e the init matrix should be of size `num_batch, number_of_units_rnn`.

The alternative is that the init matrices are always `num_batch, number_of_units_rnn` but in that case the model cannot process different sized batches and the number of parameters depends on the batch size.

craffel commented:
> you need to init with a seperate value for each sample in the batch

Hmm, I wouldn't expect that (especially when the initializations are not learned) but if there's a use case for it then so be it!  Ok, I'll ignore this for now, but it does bug me a bit.  Maybe if https://github.com/Lasagne/Lasagne/issues/11 is resolved we can change it.

skaae commented:
It is because of the arrangements of the data that you need a separate value for each sample in the batch. I have a hard time explaining it so heres an example:

If you have the words [1,2,3,4,5,6,7,8,9,10,11,12] and use a sequence length of 3 and a batch size of 2 then the data is rearranged into 2 batches:

```
batch1 = [[1,2,3],
          [7,8,9]]
batch2 = [[4,5,6]
         [10,11,12]]
```

Then you can use the last hidden state from batch1, word 3 and word 9,  to initialize the second batch which start with word 4 and word 11.

craffel commented:
> to initialize the second batch which start with word 4 and word 11.

Do you mean word 10?  If so, I think I see what you're getting at.  It seems like there should be a way to do this with a shared variable, because you can do something like that outside of the context of Lasagne, i.e. updating the initial value with the final value:

``` Python
In [1]: import theano

In [2]: import numpy as np

In [3]: init_state = theano.shared(np.array(0))

In [4]: result, updates = theano.scan(fn=lambda x: x + 1,
                                      outputs_info=init_state, n_steps=10)

In [5]: go = theano.function([], result, updates=updates)

In [6]: init_state.set_value(go()[-1])

In [7]: init_state.set_value(go()[-1])

In [8]: init_state.set_value(go()[-1])

In [9]: init_state.get_value()
Out[9]: array(30)
```

Not sure though.

skaae commented:
Yes, the numbers corresponds to words. 

You can update the shared variable, but the problem is the following lines:

``` Python
        ones = T.ones((num_batch, 1))
        if isinstance(self.cell_init, T.TensorVariable):
            cell_init = self.cell_init
        else:
            cell_init = T.dot(ones, self.cell_init)  # repeat num_batch times
```

We dot `cell_init` with `ones`. So the shared variable is only one sample but we need it to be `num_batch` samples.

f0k commented:
You'd need an additional constructor keyword argument then to distinguish between `cell_init` set to a single vector that needs to be replicated, or set to a matrix of `batch_size` vectors. (You could probably also replace the `isinstance` check with a `self.cell_init.ndim` check, but you still need to know what shape to tell `add_param()`.)

skaae commented:
I think the benefit of using a TensorVariable instead of some shared variable is that we can still properly handle `num_batch` given as None.

f0k commented:
> we can still properly handle `num_batch` given as None.

Can you? What does the `ones = T.ones((self.num_batch, 1))` line say to that? :)

skaae commented:
I posted some old code. The code is now:

https://github.com/craffel/nntools/blob/recurrent/lasagne/layers/recurrent.py#L678

``` Python

        ones = T.ones((num_batch, 1))  # num_batch is from input
        if isinstance(self.cell_init, T.TensorVariable):
            cell_init = self.cell_init
        else:
            cell_init = T.dot(ones, self.cell_init)  # repeat num_batch times

        if isinstance(self.hid_init, T.TensorVariable):
            hid_init = self.hid_init
        else:
            hid_init = T.dot(ones, self.hid_init)  # repeat num_batch times
```

It was changed with https://github.com/craffel/nntools/pull/36/

But i should probably read the code before i post it :)
